python/KAPEL.py

```diff
@@ -262,6 +262,73 @@ def process_period(config, period):
     print(f'Writing sync record to {config.output_path}/{sync_file}:')
     print('--------------------------------\n' + sync_output + '--------------------------------')
 
+def record_individual_period(config, results):
+    """ Record a record for each pod in the namespace over the summarized period.
+    Assumes each pod ran a single job and terminated upon job completion.
+    """
+    # Pivot records from {'data_type':{'pod_name':value}} to {'pod_name':{'data_type':value}}
+    per_pod_records = {}
+    for data_type, records in results.items():
+        for pod, val in records.items():
+            if not pod in per_pod_records:
+                per_pod_records[pod] = {}
+            per_pod_records[pod][data_type] = val
+    
+    dirq = QueueSimple(str(config.output_path))
+    for pod_name, records in per_pod_records.items():
+        # Only report on pods that have completed. Running pods won't have an endtime
+        if not ('starttime' in records and 'endtime' in records):
+            continue
+        individual_output = individual_message(
+            config, 
+            pod_name,
+            records.get('memory', 0),
+            records.get('cores', 0),
+            records['endtime'] - records['starttime'],
+            records.get('cpuusage', 0),
+            records['starttime'],
+            records['endtime'])
+        record_file = dirq.add(individual_output)
+        print(f'Writing individual record to {config.output_path}/{record_file}:')
+        print('--------------------------------\n' + individual_output + '--------------------------------')
+    
+
+# process a time period (do prom query, process data, write output)
+# takes a KAPELConfig object and one element of output from get_time_periods
+# Remember Prometheus queries go backwards: the time instant is the end, go backwards from there.
+def process_period(config, period):
+    period_start = period['instant'] + dateutil.relativedelta.relativedelta(seconds=-period['range_sec'])
+    print(
+        f"Processing year {period['year']}, month {period['month']}, "
+        f"querying from {period['instant'].isoformat()} and going back {period['range_sec']} s to {period_start.isoformat()}."
+    )
+    queries = QueryLogic(queryRange=(str(period['range_sec']) + 's'), namespace=config.namespace)
+
+    # SSL generally not used for Prometheus access within a cluster
+    # Docs on instant query API: https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries
+    headers = {"Authorization": config.auth_header } if config.auth_header else None
+    prom = PrometheusConnect(url=config.prometheus_server, disable_ssl=True, headers=headers)
```

Comment:
Previously the assumption was using unauthenticated HTTP inside the cluster.
But if you're passing an auth header you probably want to use HTTPS to keep it secure. 
However whether or not the URL actually specifies HTTPS is up to the configurable value set for PROMETHEUS_SERVER.
So maybe the best thing would be to [check the scheme of the URL](https://github.com/sloria/environs?tab=readme-ov-file#url-schemes) and set `disable_ssl` only if the scheme is HTTP rather than HTTPS. What do you think @mwestphall ?

Reply:
We might want to make disable_ssl configurable directly at the values.yaml level, I think its usage will vary case-by-case. It looks like it ultimately [sets `requests.Session().validate`](https://github.com/4n4nd/prometheus-api-client-python/blob/c77c0b91d85a709848e72b457ef26b706a1ceccb/prometheus_api_client/prometheus_connect.py#L77), which [only applies for https requests](https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification) and not http. In tiger, we do all our ssl termination at the ingress level so the Prometheus service would not accept https connections at the internal service address. This will probably vary cluster to cluster though.

Reply:
Good point, normally TLS happens at cluster ingress. I thought if you're authenticating inside the cluster there must be TLS inside the cluster. Maybe sending the credential in plain text inside the cluster isn't a big concern. Anyway there could be a separate issue for that if needed.